blog/2019-12-18-HowToPlanAndOptimizeASecretSanta.adoc

```diff
@@ -0,0 +1,102 @@
+[source,java]
+----
+private Constraint largerDistanceAward(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .from(SecretSantaAssignment.class)
+            .rewardConfigurableBigDecimal("secretFactor", "Larger Distance Award",
+                              (m) -> BigDecimal.valueOf(Location.calculateDistanceBetween(m.getGifter().getLocation(),
+                                                                       m.getReciever().getLocation())));
+}
+----
+
+Let try solving with just these two constraints to see what we get:
+
+image::noPairChecking.png[]
+
+Alice is assigned to Daniella, Daniella is assigned to Alice...
+Bob is assigned to Austin, Austin is assigned to Bob...
+Something about this solution seems odd; it only generate assignments
+that give gifts to one another!
+
+== What happened?
+
+This is not odd at all, but rather a consequence of how we defined
+our constraints: if Alice the person furthest from Daniella, then
+Daniella is most likely the person furthest from Alice. This means our
+"Larger Distance Award" constraint will strongly favour pairs of employees.
+
+Is this a bad thing? Yes it is; imagine if Alice got a $100 gift from
+Daniella, but only got Daniella a $10 gift. Alice will probably feel bad,
+and Daniella will feel disappointed. If Alice gave the $10 gift to Bob
+and Bob gave a $30 gift to Daniella, Alice wouldn't feel bad and
+Daniella wouldn't feel cheated.
+
+Let add this constraint (using, you guessed it, constraint streams!)
+[source,java]
+----
+private Constraint giftPair(ConstraintFactory constraintFactory) {
+    return constraintFactory
+            .fromUniquePair(SecretSantaAssignment.class,
+                            Joiners.equal(SecretSantaAssignment::getGifter, SecretSantaAssignment::getReciever),
+                            Joiners.equal(SecretSantaAssignment::getReciever, SecretSantaAssignment::getGifter))
+            .penalize("Gifter-Reciever Cycle", HardMediumSoftBigDecimalScore.ONE_MEDIUM);
+}
+----
+This constraint should be a medium constraint - we want to avoid it if possible,
+but it isn't as much of a deal breaker as someone not receiving a gift.
+
+Now let try solving again after adding the gift pair constraint:
+
+image::withPairChecking.png[]
+
+Much better; we end up with two chains: Alice gives to Austin who gives
+to Bob who gives to Charlie who gives to Alice; and Dina gives to
+Dennis who gives to Daniella who gives to Julian who gives to Dina.
+
+This is just the beginning of what you can do to optimize Secret Santa using
+OptaPlanner. For instance, you can do the following:
+
+- Allow each person to input a list of people who they do not want to be
+Secret Santa for, and add a medium constraint that ensures they are not
+the Secret Santa for anyone in their list (if possible).
+
+- Use a secret factor that sightly influences distance so people cannot
+find out who their Secret Santa is just by running OptaPlanner
```

Comment:
Should be `- Use a secret factor that slightly influences distance so people cannot
find out who their Secret Santa is just by running OptaPlanner.`